src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -605,25 +618,25 @@ private Result calculateScoreForResult(Set<ProgrammingExerciseTestCase> testCase
 
         // Case 1: There are tests and test case feedback, find out which tests were not executed or should only count to the score after the due date.
         if (!testCasesForCurrentDate.isEmpty() && !testCaseFeedback.isEmpty() && !result.getFeedbacks().isEmpty()) {
-            retainAutomaticFeedbacksWithTestCase(result, testCases);
-
-            // Copy the visibility from test case to corresponding feedback
+            filterAutomaticFeedbacksWithoutTestCase(result, testCases);
             setVisibilityForFeedbacksWithTestCase(result, testCases);
-
-            // Add feedbacks for tests that were not executed ("test was not executed").
             createFeedbackForNotExecutedTests(result, testCasesForCurrentDate);
 
-            // Add feedbacks for all duplicate test cases
-            boolean hasDuplicateTestCases = createFeedbackForDuplicateTests(result, exercise);
+            boolean hasDuplicateTestCases = createFeedbacksForDuplicateTests(result, exercise);
 
-            // Add feedback if submission penalty policy is active
             if (exercise.getSubmissionPolicy() instanceof SubmissionPenaltyPolicy penaltyPolicy) {
                 submissionPolicyService.createFeedbackForPenaltyPolicy(result, penaltyPolicy);
             }
 
             // The score is always calculated from ALL (except visibility=never) test cases, regardless of the current date!
```

Comment:
I don't see the mentioned extraction of test cases with visibility = never here. If this is expected to be done already in the input parameter, please mention this in this comment and the javadoc.

Reply:
The filter is in the method that calls this method. see `calculateScoreForResult` and `filterTestCasesForCurrentDate`